x-pack/plugin/core/src/test/java/org/elasticsearch/index/engine/FrozenIndexTests.java

```diff
@@ -0,0 +1,202 @@
+            numRefreshes++;
+            switch (randomIntBetween(0, 3)) {
+                case 0:
+                    client().prepareGet("index", "_doc", "" + randomIntBetween(0, 9)).execute(listener);
+                    break;
+                case 1:
+                    client().prepareSearch("index").setIndicesOptions(IndicesOptions.STRICT_EXPAND_OPEN_FORBID_CLOSED)
+                        .setSearchType(SearchType.QUERY_THEN_FETCH).execute(listener);
+                    // in total 4 refreshes 1x query & 1x fetch per shard (we have 2)
+                    numRefreshes += 3;
+                    break;
+                case 2:
+                    client().prepareTermVectors("index", "_doc", "" + randomIntBetween(0, 9)).execute(listener);
+                    break;
+                case 3:
+                    client().prepareExplain("index", "_doc", "" + randomIntBetween(0, 9)).setQuery(new MatchAllQueryBuilder())
+                        .execute(listener);
+                    break;
+                    default:
+                        assert false;
+            }
+        }
+        latch.await();
+        IndicesStatsResponse index = client().admin().indices().prepareStats("index").clear().setRefresh(true).get();
+        assertEquals(numRefreshes, index.getTotal().refresh.getTotal());
+    }
+
+    public void testFreezeAndUnfreeze() throws ExecutionException, InterruptedException {
+        createIndex("index", Settings.builder().put("index.number_of_shards", 2).build());
+        client().prepareIndex("index", "_doc", "1").setSource("field", "value").setRefreshPolicy(IMMEDIATE).get();
+        client().prepareIndex("index", "_doc", "2").setSource("field", "value").setRefreshPolicy(IMMEDIATE).get();
+        client().prepareIndex("index", "_doc", "3").setSource("field", "value").setRefreshPolicy(IMMEDIATE).get();
+
+        client().admin().indices().prepareFlush("index").get();
+        client().admin().indices().prepareClose("index").get();
+        XPackClient xPackClient = new XPackClient(client());
+        PlainActionFuture<AcknowledgedResponse> future = new PlainActionFuture<>();
+        TransportFreezeIndexAction.FreezeRequest request =
+            new TransportFreezeIndexAction.FreezeRequest("index");
+        xPackClient.freeze(request, future);
+        assertAcked(future.get());
+        assertAcked(client().admin().indices().prepareOpen("index"));
+        {
+            IndicesService indexServices = getInstanceFromNode(IndicesService.class);
+            Index index = resolveIndex("index");
+            IndexService indexService = indexServices.indexServiceSafe(index);
+            assertTrue(indexService.getIndexSettings().isSearchThrottled());
+            IndexShard shard = indexService.getShard(0);
+            Engine engine = IndexShardTestCase.getEngine(shard);
+            assertEquals(0, shard.refreshStats().getTotal());
+            client().admin().indices().prepareClose("index").get();
```

Comment:
nit - can you take it out of the inner scope it belongs one up (I literally popped the stack and later thought the index is still open when reading later code).